### PR TITLE
Cleanup Brave Ads startup

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Toggle ads on/off several times (in quick succession) and confirm ads service does not crash
Join Rewards and confirm users are rewarded for ads
Restart browser and confirm users are rewarded for ads

### CODE REVIEW

File: components/brave_ads/browser/ads_service_impl.cc [hunk 1]
```diff
@@ -468,197 +416,159 @@ void AdsServiceImpl::InitializePrefChangeRegistrar() {
...
+      base::BindOnce(&AdsServiceImpl::OnEnsureBaseDirectoryExists,
+                     AsWeakPtr()));
 }
 
-void AdsServiceImpl::OnEnsureBaseDirectoryExists(const uint32_t number_of_start,
-                                                 const bool success) {
+void AdsServiceImpl::OnEnsureBaseDirectoryExists(const bool success) {
   if (!success) {
     VLOG(0) << "Failed to create base directory";
     return;
```
Comment (aseren) [inline components/brave_ads/browser/ads_service_impl.cc:550]: bat ads will be in uninitialized state if we return here. Should we call Shutdown() here?

Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc]: Great catch, fixed

File: components/brave_ads/browser/ads_service_impl.cc [hunk 2]
```diff
@@ -468,197 +416,159 @@ void AdsServiceImpl::InitializePrefChangeRegistrar() {
...
-  }
-
+void AdsServiceImpl::InitializeBatAds() {
   SetSysInfo();
 
   SetBuildChannel();
 
+  GetRewardsWallet();
+
+  DCHECK(bat_ads_.is_bound());
```
[the remark this answers is not in the document]
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:563]: If the bat ads service crashes, we call `RestartBatAdsServiceAfterDelay` which calls `Shutdown`/`reset`

[the remark this answers is not in the document]
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc]: Good catch, fixed

File: components/brave_ads/browser/ads_service_impl.cc [hunk 3]
```diff
@@ -434,231 +409,172 @@ void AdsServiceImpl::OnMigrateConfirmationState(const bool success) {
     return;
```
Comment (aseren) [inline components/brave_ads/browser/ads_service_impl.cc:409]: If we failed to migrate a confirmation state, then a device id won't be retreived. Maybe we need to call GetDeviceId() even if migration failed?

**Review by aseren — APPROVED** [review]
approved with one comment

File: components/brave_ads/browser/ads_service_impl.cc (hunk 3 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:409]: If we fail confirmation, state migration ads won't run, so the device id is unnecessary. Did I miss something?

Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc:407]: Fixed. Thanks

**Pull request merged**